flink/src/test/java/org/apache/iceberg/flink/TestFlinkTableSource.java

```diff
@@ -685,4 +782,60 @@ public void testSqlParseError() {
     AssertHelpers.assertThrows("The NaN is not supported by flink now. ",
         NumberFormatException.class, () -> sql(sqlParseErrorLTE));
   }
+
+  /**
+   * The sql can be executed in both streaming and batch mode, in order to get the parallelism, we convert the flink
+   * Table to flink DataStream, so we only use streaming mode here.
+   *
+   * @throws TableNotExistException table not exist exception
+   */
+  @Test
+  public void testInferedParallelism() throws TableNotExistException {
+    Assume.assumeTrue("The execute mode should  be streaming mode", isStreamingJob);
```

Comment:
Shouldn't the inferParallelism only affect the batch job (See FlinkSource#Builder#build)?  So there's no reason that providing unit test in streaming  mode ? 

In my mind,  Providing unit tests to check whether the `inferParallelism()` is returning the expected parallelism value is enough for this changes.   Seems like The ITCase is validating the behavior of DataStreamSource#setParallelism ,  we could think it's always correct because it's a basic API in flink.

Reply:
In this way, we don't have to change so many codes in this class. Maybe we could just add unit tests in TestFlinkScan.java

Reply:
I found that in this test method,I use the flink streaming mode,but it still enter the batch mode ([here](https://github.com/apache/iceberg/blob/master/flink/src/main/java/org/apache/iceberg/flink/source/FlinkSource.java#L200)), I check the code,found that `FlinkSource.Builder#build` mthod judge streaming mode or batch mode by the conf of ScanContext instead of flink conf. will this confuse users?

Reply:
Although we use streaming mode, we read a batch of data, so flink will treat this batch job as a bounded stream. I think whether we should use `StreamExecutionEnvironment` to process batch jobs and streaming jobs at the same time?

Reply:
If someone provides both flink `Configuration` and iceberg's `properties`, then we should use the flink's `Configuration` values overwrite the iceberg's properties because  `properties` is a table-level settings  while the flink's `Configuration` is a job-level settings. It is reasonable for fine-grained configuration to ovewrite coarse-grained configuration.

Reply:
I think these are two different concepts.
In flink, whether using batch mode or streaming mode, we can read batch data. Flink treats batch jobs as bounded streaming jobs, so there should be no problem whether it is using batch mode or streaming mode to read batch data.  In addition, flink will use `StreamExecutionEnvironment (DataStream)` to do batch tasks and stream tasks uniformly ([the  doc link](https://flink.apache.org/news/2020/12/10/release-1.12.0.html#batch-execution-mode-in-the-datastream-api)) . The batch mode may expire, so I think we should also use `StreamExecutionEnvironment (DataStream)` for batch tasks as much as possible.

When we use `StreamExecutionEnvironment` , in `FlinkSource.Builder#build` method, 
the `if` and `else` block in this method are both streaming jobs, `if` block is a bounded streaming jobs, maybe we can rename `ScanContext#isStreaming` field to `ScanContext#isStreamingRead`, which will be easier to understand. `If` code block is bounded stream job (batch), `else` code block to do long-running stream job.
```
    if (!context.isStreaming()) {
        int parallelism = inferParallelism(format, context);
        return env.createInput(format, typeInfo).setParallelism(parallelism);
      } else {
        StreamingMonitorFunction function = new StreamingMonitorFunction(tableLoader, context);

        String monitorFunctionName = String.format("Iceberg table (%s) monitor", table);
        String readerOperatorName = String.format("Iceberg table (%s) reader", table);

        return env.addSource(function, monitorFunctionName)
            .transform(readerOperatorName, typeInfo, StreamingReaderOperator.factory(format));
      }
```

Reply:
> In my mind, Providing unit tests to check whether the `inferParallelism()` is returning the expected parallelism value is enough for this changes. Seems like The ITCase is validating the behavior of DataStreamSource#setParallelism , we could think it's always correct because it's a basic API in flink.

I think it’s better not to use the `inferParallelism` method to get the parallelism to do assertion, because the `inferParallelism` method is private and is an internal method of iceberg. Just as you [commented](https://github.com/apache/iceberg/pull/1936#discussion_r554983424) that it is best not to use the internal code of flink, I think we should try to use public APIs to get information.

The current `TestFlinkTableSource` class uses batch mode for unit test. In order not to modify too much code, we can move the `testInferedParallelism` method to other test classes, such as `TestFlinkScan.java`.

So I think we can use `DataStream.getTransformation().getParallelism();` to get the parallelism of the flink operator. This method is public api of flink. Even if flink is upgraded in the future, it should not be modified. What do you think?

Reply:
The `isStreaming`  indicate whether the flink source is a streaming source (In our mind) ,  not say it's a streaming job or batch job.  The hive table source also has the similar configure key : 

```java
    public static final ConfigOption<Boolean> STREAMING_SOURCE_ENABLE =
            key("streaming-source.enable")
                    .booleanType()
                    .defaultValue(false)
                    .withDescription(
                            "Enable streaming source or not.\n"
                                    + " NOTES: Please make sure that each partition/file should be written"
                                    + " atomically, otherwise the reader may get incomplete data.");
```

If we think this iceberg configure key is not very clear,  I think we could propose another separate PR to align with hive configure key.  Let's focus on this parallelism issue here,  what do you think ?

Reply:
yes,it should be a streaming source, like kafka. If necessary, we can open a separate PR to discuss this.

Reply:
> Just as you commented that it is best not to use the internal code of flink, I think we should try to use public APIs to get information.

The comment that saying we'd better not use flink's Internal API because that would introduce extra upgrade complexity (new flink version may breaks those internal API so we iceberg have to adjust the codes,  finally maintaining different versions of flink will bring us a lot of burden). 

Writing iceberg unit tests based on our iceberg's non-public ( we usually use package-access ) method is OK because there's no extra burden .

Reply:
I update the pr,move the `testInferedParallelism` method to TestFlinkScanSql,use  `FlinkSource.Builder#inferParallelism` method to do the assertion